README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggshadow

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/ggshadow)](https://CRAN.R-project.org/package=ggshadow)
[![License: GPL
v2](https://img.shields.io/badge/License-GPL_v2-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
<!-- badges: end -->

The goal of ggshadow is to provide shadow and glow geoms for points and
lines plots created with ‘ggplot2’.

## :arrow_double_down: Installation

You can install the development version of ggshadow from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
pak::pkg_install("marcmenem/ggshadow")
```

## :book: Vignette

``` r
## after installing the package
# vignette("ggshadow", package = "ggshadow")
```

## :chart: Example

### With ggshadow

``` r
library(ggplot2)
library(ggshadow)

ggplot(economics_long, aes(date, value01, colour = variable)) +
  geom_shadowline()
```

<img src="man/figures/README-example-1.png" width="100%" />

### Without ggshadow

``` r
ggplot(economics_long, aes(date, value01, colour = variable)) +
  geom_line()
```

<img src="man/figures/README-lineex-1.png" width="100%" />

### `ggshadow` supports varying the line color

``` r
ggplot(
  economics_long,
  aes(date, value01,
    group = variable,
    colour = value01,
    shadowlinewidth = 5 * (1 - value01)
  )
) +
  geom_shadowline(
    shadowcolour = "grey",
    shadowalpha = 0.5
  )
```

<img src="man/figures/README-colorvarex-1.png" width="100%" />

### `ggshadow` also provides a Neon glow style

``` r
ggplot(economics_long, aes(date, value01, color = variable)) +
  geom_glowline() +
  guides(color = "none") +
  theme(
    plot.background = element_rect(fill = "#190132"),
    panel.background = element_rect(fill = "#190132")
  )
```

<img src="man/figures/README-example-glow-1.png" width="100%" />

### Neon glow points

``` r
ggplot(mtcars, aes(wt, mpg)) +
  geom_glowpoint(color = "yellow") +
  guides(color = "none") +
  theme(
    plot.background = element_rect(fill = "#190132"),
    panel.background = element_rect(fill = "#190132")
  )
```

<img src="man/figures/README-example-glowpoint-1.png" width="100%" />

### Adding a fill below the neon glow line

``` r
ggplot(
  economics_long[economics_long$variable %in% c("pop", "unemploy"), ],
  aes(date, value01 - 0.5, color = variable, fill = variable)
) +
  geom_glowline() +
  guides(color = "none", shadowcolour = "none", fill = "none") +
  theme(
    plot.background = element_rect(fill = "#190132"),
    panel.background = element_rect(fill = "#190132")
  )
```

<img src="man/figures/README-example-glowlinefill-1.png" width="100%" />


NEWS.md
# ggshadow development

* Update `GeomShadowPath` and `GeomGlowPath` to support the linewidth parameter for ggplot2 following the [guidance from Thomas Lin Pedersen in this blog post](https://www.tidyverse.org/blog/2022/08/ggplot2-3-4-0-size-to-linewidth/).
* Remove calls to `getFromNamespace` (following the recommendation in [this blog post](https://www.tidyverse.org/blog/2022/09/playing-on-the-same-team-as-your-dependecy/) to copy code directly into the package) for non-exported functions from ggplot2.
* Add [cli](https://github.com/r-lib/cli) to Imports.
* Replace markdown README with README.Rmd file.
* Add [Eli Pousson](https://github.com/elipousson) to contributors.

# ggshadow 0.0.4

* Version 0.0.4 has a small fix to support the next version of `ggplot2`. The function `new_data_frame` was removed from `ggplot2`. It was replaced by an equivalent function from from `vctrs`.

# ggshadow 0.0.2

* Version 0.0.2 is the initial CRAN release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(ggshadow)

test_check("ggshadow")


tests/testthat/test-geom-glowpath.R
test_that("geom_glowline works", {
  p <- ggplot2::ggplot(
    ggplot2::economics_long[1:50, ],
    ggplot2::aes(date, value01, colour = variable)
  ) +
    geom_glowline()
  expect_snapshot(ggplot2::layer_data(p))
})


tests/testthat/test-geom-glowpoint.R
test_that("geom_glowpoint works", {
  p <- ggplot2::ggplot(mtcars, ggplot2::aes(wt, mpg))
  p <- p + geom_glowpoint()
  expect_snapshot(ggplot2::layer_data(p))
})


R/geom-glowpoint.r
#' Points
#'
#' The point geom is used to create scatterplots. [geom_glowpoint()] is designed
#' as a drop in replacement for [geom_point()] with an added glow beneath the
#' point to make a busy plot more aesthetically appealing or to make points
#' stand out from the rest of the plot.
#'
#' @inheritParams ggshadow-params
#' @param na.rm If `FALSE`, the default, missing values are removed with a
#'   warning. If `TRUE`, missing values are silently removed.
#' @param ... Other arguments passed on to [layer()]. These are often
#'   aesthetics, used to set an aesthetic to a fixed value, like `colour =
#'   "red"` or `size = 3`. They may also be parameters to the paired geom/stat.
#'
#'
#' @section Aesthetics:
#' Adds 3 new aesthetics to [geom_point()]:
#' * `shadowcolour` defaults to the same color as the point, controls the
#' color of the glow
#' * `shadowsize` defaults to `size`, controls the sie of the shadow.
#' * `shadowalpha` defaults to `0.06 * alpha` or `0.06`, controls
#' the alpha of the glow
#'
#' @return a layer to add to a plot.
#'
#' @export
#' @examples
#' library( ggplot2 )
#' p <- ggplot(mtcars, aes(wt, mpg))
#' p + geom_shadowpoint()
#'
geom_glowpoint <- function(mapping = NULL, data = NULL,
                       stat = "identity", position = "identity",
                       ...,
                       na.rm = FALSE,
                       show.legend = NA,
                       inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomGlowPoint,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      ...
    )
  )
}

#' @rdname ggshadow-ggproto
#' @format NULL
#' @usage NULL
#' @export
#' @importFrom stats complete.cases
#' @importFrom cli cli_warn
GeomGlowPoint <- ggproto("GeomGlowPoint", Geom,
  required_aes = c("x", "y"),
  non_missing_aes = c("size", "shape", "colour"),
  default_aes = aes(
    shape = 19, colour = "black", size = 0.3, fill = NA,
    alpha = NA, stroke = 0.5, shadowcolour=NA, shadowsize=NA, shadowalpha = NA
  ),

  handle_na = function(data, params) {
    # Drop missing values at the start or end of a line - can't drop in the
    # middle since you expect those to be shown by a break in the line
    # print( data %>% as.tbl )
    # cat('Handle NA\n')

    complete <- stats::complete.cases(data[c("x", "y", "size", "colour")])
    kept <- complete
    data <- data[kept, ]

    if (!all(kept) && !params$na.rm) {
      cli::cli_warn(
        "{.fn geom_shadowpoint}: Removed {sum(!kept)} row{?s} containing missing values."
      )
    }

    data$shadowcolour[is.na(data$shadowcolour)] <- data$colour
    data$shadowsize[is.na(data$shadowsize)] <- data$size
    data$shadowalpha[is.na(data$shadowalpha)] <- data$alpha * 0.06
    data$shadowalpha[is.na(data$shadowalpha)] <- 0.06

    data
  },

  draw_panel = function(data, panel_params, coord, na.rm = FALSE) {
    if (is.character(data$shape)) {
      data$shape <- translate_shape_string(data$shape)
    }

    coords <- coord$transform(data, panel_params)

    coords.s <- NULL
    for( i in seq(10)){
      coords.i <- coords
      coords.i$colour <- coords$shadowcolour
      coords.i$size <- coords$shadowsize * i
      coords.i$alpha <- coords$shadowalpha

      if (is.null(coords.s)){
        coords.s <- coords.i
      } else {
        coords.s <- rbind(coords.s, coords.i)
      }
    }

    # print( coords.s %>% as.tbl, n=Inf  )
    coords <- rbind( coords.s, coords )[, c( 'x', 'y', 'group', 'shape', 'colour', 'size', 'fill', 'alpha', 'stroke' )]
    # print( coords %>% as.tbl, n=Inf  )

    grid::pointsGrob(
      coords$x, coords$y,
      pch = coords$shape,
      gp = grid::gpar(
        col = scales::alpha(coords$colour, coords$alpha),
        fill = scales::alpha(coords$fill, coords$alpha),
        # Stroke is added around the outside of the point
        fontsize = coords$size * .pt + coords$stroke * .stroke / 2,
        lwd = coords$stroke * .stroke / 2
      )
    )

  }#,

  # draw_key = draw_key_point
)


R/geom-shadowpoint.r
#' Points
#'
#' The point geom is used to create scatterplots. [geom_shadowpoint()] is
#' designed as a drop in replacement for [geom_point()] with an added shadow
#' beneath the point to make a busy plot more aesthetically appealing or to make
#' points stand out from the rest of the plot.
#'
#' @inheritParams ggshadow-params
#' @param na.rm If `FALSE`, the default, missing values are removed with a
#'   warning. If `TRUE`, missing values are silently removed.
#' @param ... Other arguments passed on to [layer()]. These are often
#'   aesthetics, used to set an aesthetic to a fixed value, like `colour =
#'   "red"` or `size = 3`. They may also be parameters to the paired geom/stat.
#'
#'
#' @section Aesthetics:
#' Adds 3 new aesthetics to [geom_point()]:
#' * `shadowcolour` defaults to white, controls the color of the shadow.
#' * `shadowsize` defaults to `1.8 * size`, controls the size of the
#' shadow.
#' * `shadowalpha` defaults to `0.25 * alpha` or `0.9`, controls
#' the alpha of the shadow.
#'
#' @return a layer to add to a plot.
#'
#' @export
#' @examples
#' library( ggplot2 )
#' p <- ggplot(mtcars, aes(wt, mpg))
#' p + geom_shadowpoint()
#'
geom_shadowpoint <- function(mapping = NULL, data = NULL,
                       stat = "identity", position = "identity",
                       ...,
                       na.rm = FALSE,
                       show.legend = NA,
                       inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomShadowPoint,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      ...
    )
  )
}

#' @rdname ggshadow-ggproto
#' @format NULL
#' @usage NULL
#' @export
#' @importFrom cli cli_warn
#' @importFrom stats complete.cases
#' @importFrom grid pointsGrob gpar
GeomShadowPoint <- ggproto("GeomShadowPoint", Geom,
  required_aes = c("x", "y"),
  non_missing_aes = c("size", "shape", "colour"),
  default_aes = aes(
    shape = 19, colour = "black", size = 1.5, fill = NA,
    alpha = NA, stroke = 0.5, shadowcolour=NA, shadowsize=NA, shadowalpha = NA
  ),

  handle_na = function(data, params) {
    # Drop missing values at the start or end of a line - can't drop in the
    # middle since you expect those to be shown by a break in the line
    # print( data %>% as.tbl )
    # cat('Handle NA\n')

    complete <- stats::complete.cases(data[c("x", "y", "size", "colour")])
    kept <- complete
    data <- data[kept, ]

    if (!all(kept) && !params$na.rm) {
      cli::cli_warn(
        "{.fn geom_shadowpoint}: Removed {sum(!kept)} row{?s} containing missing values."
      )
    }

    data$shadowcolour[is.na(data$shadowcolour)] <- 'white'
    data$shadowsize[is.na(data$shadowsize)] <- data$size * 1.8
    data$shadowalpha[is.na(data$shadowalpha)] <- data$alpha * 0.25
    data$shadowalpha[is.na(data$shadowalpha)] <- 0.9

    data
  },

  draw_panel = function(data, panel_params, coord, na.rm = FALSE) {
    if (is.character(data$shape)) {
      data$shape <- translate_shape_string(data$shape)
    }

    coords <- coord$transform(data, panel_params)

    # print( coords %>% as.tbl, n=Inf  )

    coords.s <- coords
    coords.s$colour <- coords.s$shadowcolour
    coords.s$size <- coords.s$shadowsize
    coords.s$alpha <- coords.s$shadowalpha

    # print( coords.s %>% as.tbl, n=Inf  )

    coords <- rbind( coords.s, coords )[, c( 'x', 'y', 'group', 'shape', 'colour', 'size', 'fill', 'alpha', 'stroke' )]

    # print( coords %>% as.tbl, n=Inf  )
    # print( alpha(coords$fill, ifelse( is.na(coords$fill), NA, coords$shadowcolour)) )

    grid::pointsGrob(
      coords$x, coords$y,
      pch = coords$shape,
      gp = grid::gpar(
        col = alpha(coords$colour, coords$alpha),
        fill = alpha(coords$fill, coords$alpha),
        # Stroke is added around the outside of the point
        fontsize = coords$size * .pt + coords$stroke * .stroke / 2,
        lwd = coords$stroke * .stroke / 2
      )
    )

  }
)


R/geom-glowpath.r


#' Connect Observations
#'
#' Plot a glow beneath the connected lines to make it easier to read a chart with several
#' overlapping observations. `geom_glowpath()` connects the observations in the order in which they appear
#' in the data. `geom_glowline()` connects them in order of the variable on the
#' x axis. `geom_glowstep()` creates a stairstep plot, highlighting exactly
#' when changes occur.
#'
#' The `group` aesthetic determines which cases are
#' connected together. These functions are designed as a straight replacement
#' to the [geom_path()], [geom_line()] and [geom_step()] functions.
#' To set the order of drawing, make the `colour` aesthetic a factor, and set the order
#' from bottom to top.
#'
#' @inheritParams ggshadow-params
#' @param lineend Line end style (round, butt, square).
#' @param linejoin Line join style (round, mitre, bevel).
#' @param linemitre Line mitre limit (number greater than 1).
#' @param arrow Arrow specification, as created by [grid::arrow()].
#' @param na.rm If `FALSE`, the default, missing values are removed with
#'   a warning. If `TRUE`, missing values are silently removed.
#' @param ... Other arguments passed on to [layer()]. These are
#'   often aesthetics, used to set an aesthetic to a fixed value, like
#'   `colour = "red"` or `linewidth = 3`. They may also be parameters
#'   to the paired geom/stat.
#' @param orientation The orientation of the layer. The default (`NA`)
#' automatically determines the orientation from the aesthetic mapping. In the
#' rare event that this fails it can be given explicitly by setting `orientation`
#' to either `"x"` or `"y"`. See the *Orientation* section for more detail.
#'
#'
#' @seealso
#'  [ggplot2::geom_path()], [ggplot2::geom_line()], [ggplot2::geom_step()]:
#'  Filled paths (polygons);
#'
#' @section Missing value handling:
#' `geom_glowpath()`, `geom_glowline()`, and `geom_glowstep()` handle `NA` as
#' follows:
#'
#' *  If an `NA` occurs in the middle of a line, it breaks the line. No warning
#'    is shown, regardless of whether `na.rm` is `TRUE` or `FALSE`.
#' *  If an `NA` occurs at the start or the end of the line and `na.rm` is `FALSE`
#'    (default), the `NA` is removed with a warning.
#' *  If an `NA` occurs at the start or the end of the line and `na.rm` is `TRUE`,
#'    the `NA` is removed silently, without warning.
#'
#' @section Aesthetics:
#' Adds 3 new aesthetics to [geom_path()]:
#' * `shadowcolour` defaults to path color, controls the color of the shadow.
#' * `shadowlinewidth` defaults to `linewidth`, controls the linewidth of the shadow.
#' * `shadowalpha` defaults to `0.06 * alpha` or `0.06`, controls the alpha of the glow.
#'
#' @return a `ggplot2` layer to add to a plot.
#'
#' @export
#' @examples
#' # geom_glowline() is suitable for time series
#' library(ggplot2)
#' ggplot(economics_long, aes(date, value01, colour = variable)) + geom_glowline()
#'
#'
#' @describeIn geom_glowpath Connects observations in the order in which they appear in the data.
geom_glowpath <- function(mapping = NULL, data = NULL,
                      stat = "identity", position = "identity",
                      ...,
                      lineend = "butt",
                      linejoin = "round",
                      linemitre = 10,
                      arrow = NULL,
                      na.rm = FALSE,
                      show.legend = NA,
                      inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomGlowPath,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      lineend = lineend,
      linejoin = linejoin,
      linemitre = linemitre,
      arrow = arrow,
      na.rm = na.rm,
      ...
    )
  )
}

#' @rdname ggshadow-ggproto
#' @format NULL
#' @usage NULL
#' @export
#' @importFrom cli cli_warn cli_inform
#' @importFrom grid gList gpar segmentsGrob polylineGrob polygonGrob
#' @importFrom stats complete.cases ave
#' @importFrom scales alpha
#' @importFrom ggplot2 draw_key_path
#' @importFrom vctrs new_data_frame
GeomGlowPath <- ggproto("GeomGlowPath", Geom,
                    required_aes = c("x", "y"),

                    default_aes = aes(colour = "black", linewidth = 0.5, linetype = 1, alpha = NA,
                                      shadowcolour=NA, shadowlinewidth=NA, shadowalpha = NA,
                                      fill=NA),

                    setup_data = function(data, params) {
                      data$flipped_aes <- params$flipped_aes
                      data <- flip_data(data, params$flipped_aes)
                      data <- transform(data[order(data$PANEL, data$group), ], ymin = 0, ymax = y)

                      # cat( crayon::yellow('setup_data (GlowPath)\n') )
                      # print( data %>% as_tibble )

                      flip_data(data, params$flipped_aes)
                    },

                    handle_na = function(data, params) {
                      # Drop missing values at the start or end of a line - can't drop in the
                      # middle since you expect those to be shown by a break in the line
                      # print( data %>% as.tbl )
                      # cat('Handle NA\n')

                      complete <- stats::complete.cases(data[c("x", "y", "linewidth", "colour", "linetype")])
                      kept <- stats::ave(complete, data$group, FUN = keep_mid_true)
                      data <- data[kept, ]

                      if (!all(kept) && !params$na.rm) {
                        cli::cli_warn(
                          "{.fn geom_glowpath}: Removed {sum(!kept)} row{?s} containing missing values."
                        )
                      }

                      data$shadowcolour[is.na(data$shadowcolour)] <- data$colour
                      data$linewidth <- data$linewidth %||% data$size
                      data$shadowlinewidth[is.na(data$shadowlinewidth)] <- data$linewidth
                      data$shadowalpha[is.na(data$shadowalpha)] <- data$alpha * 0.06
                      data$shadowalpha[is.na(data$shadowalpha)] <- 0.06

                      # cat( crayon::yellow('handle_na (GlowPath)\n') )
                      # print( data %>% as_tibble )

                      data
                    },

                    draw_panel = function(data, panel_params, coord, arrow = NULL,
                                          lineend = "butt", linejoin = "round", linemitre = 10,
                                          na.rm = FALSE) {
                      if (!anyDuplicated(data$group)) {
                        cli::cli_inform(
                          c(
                            "!" = "{.fn geom_glowpath}: Each group consists of
                            only one observation.",
                            "*" = "Do you need to adjust the group aesthetic?"
                          )
                        )
                      }

                      # cat( crayon::yellow('draw_panel data (GlowPath)\n') )
                      # print( data %>% as_tibble )

                      data$shadowlinewidth <- data$shadowlinewidth %||% data$shadowsize

                      # must be sorted on group
                      data <- data[order(data$group), , drop = FALSE]
                      munched <- coord_munch(coord, data, panel_params)

                      # cat( crayon::yellow('draw_panel munched (GlowPath)\n') )
                      # print( munched %>% as_tibble )

                      # Silently drop lines with less than two points, preserving order
                      rows <- stats::ave(seq_len(nrow(munched)), munched$group, FUN = length)
                      munched <- munched[rows >= 2, ]
                      if (nrow(munched) < 2) return(ggplot2::zeroGrob())

                      # Work out whether we should use lines or segments
                      attr <- dapply(munched, "group", function(df) {
                        linetype <- unique(df$linetype)
                        vctrs::new_data_frame(
                          solid = identical(linetype, 1) || identical(linetype, "solid"),
                          constant = nrow(unique(df[, c("alpha", "colour","linewidth", "linetype", 'shadowcolour', 'shadowlinewidth', 'shadowalpha', 'fill')])) == 1
                        )
                      })
                      solid_lines <- all(attr$solid)
                      constant <- all(attr$constant)
                      if (!solid_lines && !constant) {
                        cli::cli_abort(
                          "{.fn geom_glowpath}: {.arg {c('colour', 'linewidth', 'linetype')}} must be constant over the
                          line when {.arg linetype} is not 'solid'."
                        )
                      }

                      # Work out grouping variables for grobs
                      n <- nrow(munched)
                      group_diff <- munched$group[-1] != munched$group[-n]
                      start <- c(TRUE, group_diff)
                      end <-   c(group_diff, TRUE)

                      if (!constant) {
                        cli::cli_inform(
                          c(
                            "i" = "Varying color, alpha, linewidth, linetype or shadow values are
                            not supported for {.fn GeomGlowPath}.",
                            ">" = "Defaulting to {.fn GeomShadowLine}"
                          )
                        )
                        #print( munched %>% as.tbl, n=150 )
                        munched$start <- start
                        munched$end <- end

                        munched.s <- munched
                        munched.s$shadow <- T
                        munched.s$colour <- munched.s$shadowcolour
                        munched.s$linewidth <- munched.s$shadowlinewidth
                        munched.s$alpha <- munched.s$shadowalpha

                        munched$shadow <- F

                        munched <- rbind( munched.s, munched)
                        munched$id <- 2*match(munched$group, unique(munched$group)) - munched$shadow

                        munched <- munched[order(munched$group), c('colour', 'linewidth', 'y', 'x', 'linetype','alpha', 'id', 'start', 'end')]

                        aph <- scales::alpha( munched$colour[munched$start], munched$alpha[munched$start] )

                        # print( munched %>% as.tbl, n=50 )

                        grid::segmentsGrob(
                          munched$x[!munched$end],
                          munched$y[!munched$end],
                          munched$x[!munched$start],
                          munched$y[!munched$start],
                          default.units = "native", arrow = arrow,
                          gp = grid::gpar(
                            col = scales::alpha(munched$colour, munched$alpha)[!munched$end],
                            fill = scales::alpha(munched$colour, munched$alpha)[!munched$end],
                            lwd = munched$linewidth[!munched$end] * .pt,
                            lty = munched$linetype[!munched$end],
                            lineend = lineend,
                            linejoin = linejoin,
                            linemitre = linemitre
                          )
                        )
                      } else {
                        # print( munched %>% as.tbl, n=15 )
                        munched$start <- start
                        munched$layernum <- 0
                        munched.i <- munched

                        for( i in seq(10)){
                          munched.s <- munched.i
                          munched.s$layernum <- i
                          munched.s$colour <- munched.s$shadowcolour
                          munched.s$linewidth <- (munched.s$shadowlinewidth %||% munched.s$shadowsize) * i
                          munched.s$alpha <- munched.s$shadowalpha

                          munched <- rbind( munched.s, munched)
                        }
                        munched$id <- 11*match(munched$group, unique(munched$group)) - munched$layernum

                        # print( munched %>% as_tibble, n=150 )
                        munched <- munched[order(munched$group), c('colour', 'linewidth', 'y', 'x', 'linetype','alpha', 'id', 'start')]

                        aph <- scales::alpha( munched$colour[munched$start], munched$alpha[munched$start] )

                        g_lines <- grid::polylineGrob(
                          munched$x, munched$y, id = munched$id,
                          default.units = "native", arrow = arrow,
                          gp = gpar(
                            col = aph,
                            fill = aph,
                            lwd = munched$linewidth[munched$start] * .pt,
                            lty = munched$linetype[munched$start],
                            lineend = lineend,
                            linejoin = linejoin,
                            linemitre = linemitre
                          )
                        )


                        if (!is.na(munched.i$fill[1])){
                          grl <- unique(munched.i$group)

                          polys <- grid::gList(g_lines)
                          for( gr in grl){
                            munched.g <- subset(munched.i, group == gr)
                            fillcolour <- munched.g$fill[1]
                            fillalpha  <- munched.g$shadowalpha[1] * 8

                            # cat(crayon::red('Group', gr), fillcolour, fillalpha, '\n')
                            # print( munched.g %>% as_tibble) #, n=Inf )

                            minx <- min( munched.g$x )
                            maxx <- max( munched.g$x )
                            miny <- munched.g$ymin[1]

                            g_poly <- grid::polygonGrob(
                              c(minx, munched.g$x, maxx),
                              c(miny, munched.g$y, miny),
                              default.units = "native",
                              gp = grid::gpar(
                                fill = scales::alpha(fillcolour, fillalpha),
                                col = NA
                              )
                            )

                            polys <- grid::gList( polys, g_poly )
                          }

                          # str( polys )

                          ggname("geom_glowline", grid::grobTree(polys))
                        } else {
                          # cat(crayon::red('No fill colour\n'))
                          ggname("geom_glowline", g_lines)
                        }

                      }
                    },

                    draw_key = ggplot2::draw_key_path,

                    non_missing_aes = "size",
                    rename_size = TRUE
)



#' @describeIn geom_glowpath Connects observations in order of the variable on the x axis.
#' @export
geom_glowline <- function(mapping = NULL, data = NULL, stat = "identity",
                      position = "identity", na.rm = FALSE, orientation = NA,
                      show.legend = NA, inherit.aes = TRUE, ...) {
  # cat('glow line\n')

  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomGlowLine,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      orientation = orientation,
      ...
    )
  )
}

#' @rdname ggshadow-ggproto
#' @format NULL
#' @usage NULL
#' @export
#' @include geom-glowpath.r
GeomGlowLine <- ggproto("GeomGlowLine", GeomGlowPath,
                    setup_params = function(data, params) {
                      params$flipped_aes <- has_flipped_aes(data, params, ambiguous = TRUE)
                      params
                    },

                    extra_params = c("na.rm", "orientation"),

                    setup_data = function(data, params) {
                      data$flipped_aes <- params$flipped_aes
                      data <- flip_data(data, params$flipped_aes)

                      #data <- data[order(data$PANEL, data$group, data$x), ]
                      data <- transform(data[order(data$PANEL, data$group, data$x), ], ymin = 0, ymax = y)

                      # cat( crayon::yellow('setup_data (GlowLine)\n') )
                      # print( data %>% as_tibble )

                      flip_data(data, params$flipped_aes)
                    }
)

#' @describeIn geom_glowpath Creates a stairstep plot, highlighting exactly when changes occur.
#' @param direction direction of stairs: 'vh' for vertical then horizontal,
#'   'hv' for horizontal then vertical, or 'mid' for step half-way between
#'   adjacent x-values.
#' @export
geom_glowstep <- function(mapping = NULL, data = NULL, stat = "identity",
                      position = "identity", direction = "hv",
                      na.rm = FALSE, show.legend = NA, inherit.aes = TRUE, ...) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomGlowStep,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      direction = direction,
      na.rm = na.rm,
      ...
    )
  )
}

#' @rdname ggshadow-ggproto
#' @format NULL
#' @usage NULL
#' @export
#' @include geom-glowpath.r
GeomGlowStep <- ggproto("GeomGlowStep", GeomGlowPath,
                    draw_panel = function(data, panel_params, coord, direction = "hv") {
                      data <- dapply(data, "group", stairstep, direction = direction)
                      GeomGlowPath$draw_panel(data, panel_params, coord)
                    }
)
